ibis/pandas/execution/timecontext.py

```diff
@@ -0,0 +1,90 @@
+""" Implementation of compute_time_context for time context related operations
+
+Time context of a node is computed at the beginning of execution phase.
+
+To use time context to load time series data:
+
+For operations like window, asof_join that adjust time context in execution,
+implement ``compute_time_context`` to pass different time contexts to child
+nodes.
+
+If ``pre_execute`` preloads any data, it should use timecontext to trim data
+to be in the time range.
+
+``execute_node`` of a leaf node can use timecontext to trim data, or to pass
+it as a filter in the database query.
+
+In some cases, data need to be trimmed in ``post_execute``.
+
+Note: In order to use the feature we implemented here, there must be a
+column of Timestamp type, and named as 'time' in TableExpr. And this 'time'
+column should be preserved across the expression tree. If 'time' column is
+dropped then execution will result in error.
+See ``execute_database_table_client`` in ``generic.py``.
+And we assume timecontext is passed in as a tuple (begin, end) where begin and
+end are timestamp, or datetime string like "20100101".
+
+This is an optional feature. The result of executing an expression without time
+context is conceptually the same as executing an expression with (-inf, inf)
+time context.
+"""
+
+import pandas as pd
+
+import ibis.expr.api as ir
+import ibis.expr.operations as ops
+from ibis.pandas.core import compute_time_context
+from ibis.pandas.execution import execute
+
+
+@compute_time_context.register(ops.AsOfJoin)
+def adjust_context_asof_join(op, num_args, timecontext, **kwargs):
```

Comment:
Discussed off line. It's a little cleaner to remove `num_args` and compute `num_args` inside.

Reply:
Yes, passing `num_args` or list of `computable_args` are redundant to the api and it doesn't hurt performance much. The developer that extend this function should handle args and compute in their dispatch implementation.